update FA documentation (#473)
FirelyTeam/firely-docs-firely-server: security/firely-auth/firely-auth-management-app.rst

Firely Auth User Management
===========================

.. note:: 

>   In Firely Auth versions < 4.0.0 there was an executable to manage users, this has been deprecated.

Reviewer: ```suggestion
  In Firely Auth versions < 4.0.0 there was an executable to manage users, this has been deprecated and removed.
```

Author: done

After you have installed, started and created your admin account, you can now manage users in the UI.

> After creation of a non-sso account, an email will be sent to the address of the account to activate the account. When this link would be expired, you can resend this email through the UI as well.

Reviewer: ```suggestion
After creation of a local account (not using Single Sign-on), an email will be sent to the address of the account to activate the account. When this link expires, an admin can resend this email through the UI.
```

Author: done

> When you want to manage the users programmatically, you can use the API's that are specified in the swagger documentation which you can find at: https://localhost:5001/swagger/ (or replace localhost with the url of your Firely Auth installation).

Reviewer: ```suggestion
When you want to manage the users programmatically, you can use the API's that are specified in the swagger documentation which you can find at: https://localhost:5001/swagger/ (or replace localhost with the url of your Firely Auth installation). The API enables the creation, updating and deletion of users.
```

Author: done

> When you want to manage the users programmatically, you can use the API's that are specified in the swagger documentation which you can find at: https://localhost:5001/swagger/ (or replace localhost with the url of your Firely Auth installation).

Reviewer: Could we add some examples here with the payload on how to call the service?

To be able to use the api, you will need a client token (obtained trough the client_credentials flow) for a client who has the ``AllowManagementApiAccess`` property set to true.

> It is recommended to use a specific client for this API access.

Reviewer: How about adding a default appsetting here which can be added via copy&paste?

Author: done

You can set the passwordHashIterations in an environment variable ``FIRELY_IDENTITY_SERVER_USERS_PASSWORD_HASH_ITERATIONS`` so it doesn't need repeating.

- ``./Firely.IdentityServer.ManagementApp.exe```

  - ``-?, -h, --help``: Show available commands
  - ``users``: invoke user management

    - ``list``: list all users
      
      ``-cs, --connectionString``
      
      ``-hi, --passwordHashIterations``

    - ``create``: create a new user, specifying username and password
      
      ``-u, --username``: username
      
      ``-p, --password``: password - enclose in double quotes if it contains spaces.
      
      ``-cl`, --claim <name=value> [<name=value> ...]``: list of additional claims for this user, see 'AdditionalClaims' under :ref:`firely_auth_settings_userstore`
      
      ``-cs, --connectionString``
      
      ``-hi, --passwordHashIterations``

    - ``delete``: delete a user so the user no longer has access
      
      ``-u, --username``: username
      
      ``-cs, --connectionString``
      
      ``-hi, --passwordHashIterations``

.. _firely_auth_mgmt_password:

> Password requirements

Reviewer: Do we have password requirements listed somewhere else?

Author: in the app settings

Examples
--------

For all examples you need to provide a connectionString. It is easiest to set this upfront in an environment variable:

.. code-block:: powershell

    $env:FIRELY_IDENTITY_SERVER_USERS_CONNECTION_STRING = "Server=localhost,1433;MultipleActiveResultSets=true;Database=firely_auth_store;User Id=<db_user>;Password=<db_user_password>;Encrypt=True"

List the users in the database
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: powershell

    ./Firely.IdentityServer.ManagementApp.exe users list

Create a new user in the database
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: powershell

    ./Firely.IdentityServer.ManagementApp.exe users create -u bob -p 1P@ssword -c patient=Patient/bob

Try to issue the same command again to see that a user with the same name is rejected.

Delete a user from the database
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: powershell

    ./Firely.IdentityServer.ManagementApp.exe users delete -u bob

Change a user's password
^^^^^^^^^^^^^^^^^^^^^^^^

There is no function to change a user's password directly. Instead delete the user and recreate it with the new password and any claims.

.. _firely_auth_mgmt_sso_user:

> Creating password-less users in the UI for SSO

Author: yeah, API or UI

> Creating password-less users in the UI for SSO

Author: just mark them as SSO user
